redaxo/src/core/lib/sql/table.php

```diff
@@ -308,6 +340,125 @@ public function setPrimaryKey($columns)
     }
 
     /**
+     * @param string $name
+     *
+     * @return bool
+     */
+    public function hasIndex($name)
+    {
+        return isset($this->indexes[$name]);
+    }
+
+    /**
+     * @param string $name
+     *
+     * @return null|rex_sql_index
+     */
+    public function getIndex($name)
+    {
+        if (!$this->hasIndex($name)) {
+            return null;
+        }
+
+        return $this->indexes[$name];
+    }
+
+    /**
+     * @return rex_sql_index[]
+     */
+    public function getIndexes()
+    {
+        return $this->indexes;
+    }
+
+    /**
+     * @param rex_sql_index $index
+     *
+     * @return $this
+     */
+    public function addIndex(rex_sql_index $index)
+    {
+        $name = $index->getName();
+
+        if ($this->hasIndex($name)) {
+            throw new RuntimeException(sprintf('Index "%s" already exists.', $name));
+        }
+
+        $this->indexes[$name] = $index;
+
+        return $this;
+    }
+
+    /**
+     * @param rex_sql_index $index
+     *
+     * @return $this
+     */
+    public function ensureIndex(rex_sql_index $index)
+    {
+        $name = $index->getName();
+
+        if (!$this->hasIndex($name)) {
+            return $this->addIndex($index);
+        }
+
+        if ($this->getIndex($name)->equals($index)) {
+            return $this;
+        }
+
+        $this->indexes[$name] = $index->setModified(true);
+
+        return $this;
+    }
+
+    /**
+     * @param string $oldName
+     * @param string $newName
+     *
+     * @return $this
+     *
+     * @throws rex_exception
+     */
+    public function renameIndex($oldName, $newName)
+    {
+        if (!$this->hasIndex($oldName)) {
+            throw new rex_exception(sprintf('Index with name "%s" does not exist.', $oldName));
```

Comment:
Haben wir eine rex_sql_exception?

Reply:
Haben wir, bin mir aber nicht sicher, ob die nicht nur für wirkliche sql-operationen verwendet werden sollte, also in `rex_sql`. Also wenn Queries auf Sql-Ebene fehlschlagen etc.
Hier ist ja ein Vorabtest.
Aber keine Ahnung.